router/widen-lemma.gobra

```diff
@@ -86,39 +107,26 @@ requires  0 <= length && length <= len(raw)
 requires dp.Valid()
 requires acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), R50)
 requires acc(sl.AbsSlice_Bytes(raw[:length], 0, len(raw[:length])), R50)
-requires validPktMetaHdr(raw)
-requires validPktMetaHdr(raw[:length])
+requires slayers.ValidPktMetaHdr(raw)
+requires slayers.ValidPktMetaHdr(raw[:length])
 ensures  acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), R50)
 ensures  acc(sl.AbsSlice_Bytes(raw[:length], 0, len(raw[:length])), R50)
-ensures  validPktMetaHdr(raw)
-ensures  validPktMetaHdr(raw[:length])
+ensures  slayers.ValidPktMetaHdr(raw)
+ensures  slayers.ValidPktMetaHdr(raw[:length])
 ensures  absPkt(dp, raw) == absPkt(dp, raw[:length])
 decreases
 func absPktWidenLemma(dp io.DataPlaneSpec, raw []byte, length int) {
 
-	// declarations
-	var last1 io.IO_as
-	var last2 io.IO_as
-	var first1 io.IO_as
-	var first2 io.IO_as
-	var leftAsidSeq1 option[seq[io.IO_as]]
-	var leftAsidSeq2 option[seq[io.IO_as]]
-	var rightAsidSeq1 option[seq[io.IO_as]]
-	var rightAsidSeq2 option[seq[io.IO_as]]
-	var midAsidSeq1 option[seq[io.IO_as]]
-	var midAsidSeq2 option[seq[io.IO_as]]
-	var midAsid1 option[io.IO_as]
-	var midAsid2 option[io.IO_as]
-	var ret1 option[io.IO_pkt2]
-	var ret2 option[io.IO_pkt2]
-	var lm bool
-
-	// abspkt step by step
-	_ := reveal validPktMetaHdr(raw)
-	_ := reveal validPktMetaHdr(raw[:length])
-	hdr1 := unfolding acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), R51) in binary.BigEndian.Uint32(raw[0:4])
-	hdr2 := unfolding acc(sl.AbsSlice_Bytes(raw[:length], 0, len(raw[:length])), R51) in binary.BigEndian.Uint32(raw[:length][0:4])
-	assert unfolding acc(sl.AbsSlice_Bytes(raw[:length], 0, len(raw[:length])), R51) in binary.BigEndian.Uint32(raw[:length][0:4]) == unfolding acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), R51) in binary.BigEndian.Uint32(raw[:length][0:4])
+	_ := reveal slayers.ValidPktMetaHdr(raw)
+	_ := reveal slayers.ValidPktMetaHdr(raw[:length])
+	headerOffset1 := slayers.GetAddressOffset(raw)
+	headerOffset2 := slayers.GetAddressOffset(raw[:length])
+
+	assert forall k int :: {&raw[headerOffset1:headerOffset1+scion.MetaLen][k]} 0 <= k && k < scion.MetaLen ==> &raw[headerOffset1:headerOffset1+scion.MetaLen][k] == &raw[headerOffset1 + k]
+	assert forall k int :: {&raw[headerOffset2:headerOffset2+scion.MetaLen][k]} 0 <= k && k < scion.MetaLen ==> &raw[headerOffset2:headerOffset2+scion.MetaLen][k] == &raw[headerOffset2 + k]
+	hdr1 := unfolding acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), R51) in binary.BigEndian.Uint32(raw[headerOffset1:headerOffset1+scion.MetaLen])
+	hdr2 := unfolding acc(sl.AbsSlice_Bytes(raw[:length], 0, len(raw[:length])), R51) in binary.BigEndian.Uint32(raw[:length][headerOffset2:headerOffset2+scion.MetaLen])
+	assert unfolding acc(sl.AbsSlice_Bytes(raw[:length], 0, len(raw[:length])), R51) in binary.BigEndian.Uint32(raw[:length][headerOffset2:headerOffset2+scion.MetaLen]) == unfolding acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), R51) in binary.BigEndian.Uint32(raw[headerOffset1:headerOffset1+scion.MetaLen])
```

Comment:
these lines are very hard to read. for the first two assertions, I would introduce a line break after the `==>`. Also, I think that Viper ends up ignoring these triggers because they have arithmetic operations. Maybe a better alternative for the first quantifier would be:
```go
assert forall k int :: {&raw[k]} headerOffset1 <= k && k < headerOffset1 + scion.MetaLen ==>
    &raw[headerOffset1:headerOffset1+scion.MetaLen][k-headerOffset1] == &raw[k]
```
Similar for the second quantifier.

For the definition of `hdr1` and `hdr2`, it would help if we had a line break after `in`.

For the last quantifier, why don't we use `hdr1` and `hdr2` instead of repeating the large expressions? (unless I am reading this wrongly, which I might xD)